core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala

```diff
@@ -0,0 +1,220 @@
+            if (useStd3ASCIIRules) {
+              loop(acc.appendCodePoint(ReplacementCharacter), MappingError(index, "Disallowed code point in input.", CodePoint.unsafeFromInt(value)) :: errors, index + indexIncrement)
+            } else {
+              loop(unsafeLookupDisallowedSTD3CodePointMultiMapping0(value).foldLeft(acc){
+                case (acc, value) =>
+                  acc.appendCodePoint(value)
+              }, errors, index + indexIncrement)
+            }
+          case -12 => // DEVIATION_IGNORED
+            loop(acc, errors, index + indexIncrement)
+          case otherwise => // MAPPED or bug
+            if (otherwise < 0) {
+              throw new AssertionError(s"Unexpected mapping result (this is probably a bug in idna4s): $otherwise")
+            } else {
+              // This is a mapped code point
+              loop(acc.appendCodePoint(otherwise), errors, index + indexIncrement)
+            }
+        }
+      }
+
+    loop(new StringBuilder(len), Nil, 0)
+  }
+
+  def mapCodePoint(codePoint: CodePoint): CodePointStatus = {
+    import CodePointStatus._
+
+    (unsafeMapIntCodePointSentinel0(codePoint.value): @switch) match {
+      // Literals used rather than named constants because scalac won't
+      // generate a switch with the named constants.
+      case -1 => // VALID
+          Valid.always
+      case -2 => // VALID_NV8
+        Valid.nv8
+      case -3 => // VALID_XV8
+        Valid.xv8
+      case -4 => // IGNORED
+        Ignored.instance
+      case -5 => // MAPPED_MULTI_CODE_POINT
+        Mapped.of(unsafeLookupCodePointMultiMapping0(codePoint.value).map(CodePoint.unsafeFromInt))
+      case -6 => // DEVIATION
+        Deviation.one(CodePoint.unsafeFromInt(unsafeLookupDeviationCodePointMapping0(codePoint.value)))
+      case -7 => // DEVIATION_MULTI
+        Deviation.of(unsafeLookupDeviationCodePointMultiMapping0(codePoint.value).map(CodePoint.unsafeFromInt).toList)
+      case -8 => // DISALLOWED
+        Disallowed.instance
+      case -9 => // DISALLOWED_STD3_VALID
+        Disallowed_STD3_Valid.instance
+      case -10 => // DISALLOWED_STD3_MAPPED
+        Disallowed_STD3_Mapped.one(CodePoint.unsafeFromInt(unsafeLookupDisallowedSTD3CodePointMapping0(codePoint.value)))
+      case -11 => // DISALLOWED_STD3_MAPPED_MULTI
+        Disallowed_STD3_Mapped.of(unsafeLookupDisallowedSTD3CodePointMultiMapping0(codePoint.value).map(CodePoint.unsafeFromInt))
+      case -12 => // DEVIATION_IGNORED
+        Deviation.ignored
+      case otherwise =>
+        // Mapped. There is no sentinel value for mapped, because in this case
+        // a code point maps to a single new code point, we just return the
+        // mapping directly.
+        Mapped.one(CodePoint.unsafeFromInt(otherwise))
+    }
+  }
+
+  def unsafeMapIntCodePoint(codePoint: Int): Either[String, CodePointStatus] =
+    CodePoint.fromInt(codePoint).map(mapCodePoint)
+
+  private val ReplacementCharacter: Int =
+    0xFFFD
```

Comment:
This allows it to be inlined at use, both the `final` and lack of type annotation are necessary.

https://www.scala-lang.org/files/archive/spec/2.13/04-basic-declarations-and-definitions.html#value-declarations-and-definitions

```suggestion
  private final val ReplacementCharacter = 0xFFFD
```

Reply:
TIL. I'll update that.